compiler/src/dotty/tools/dotc/core/unpickleScala2/Scala2Erasure.scala

```diff
@@ -0,0 +1,261 @@
+      supportedType(tp2)
+    case RefinedType(parent, _, _) =>
+      supportedType(parent)
+    case AnnotatedType(parent, _) if parent.dealias.isInstanceOf[Scala2RefinedType] =>
+      throw new TypeError(i"Unsupported Scala 2 type: Component $parent of intersection is annotated.")
+    case tp @ TypeRef(prefix, _) if !tp.symbol.exists && prefix.dealias.isInstanceOf[Scala2RefinedType] =>
+      throw new TypeError(i"Unsupported Scala 2 type: Prefix $prefix of intersection component is an intersection or refinement.")
+    case _ =>
+
+  /** A type that would be represented as a RefinedType in Scala 2.
+   *
+   *  The `RefinedType` of Scala 2 contains both a list of parents
+   *  and a list of refinements, intersections are represented as a RefinedType
+   *  with no refinements.
+   */
+  type Scala2RefinedType = RefinedType | AndType
+
+  /** A TypeRef that is known to represent a member of a structural type. */
+  type StructuralRef = TypeRef
+
+  /** The equivalent of a Scala 2 type symbol.
+   *
+   *  In some situations, nsc will create a symbol for a type where we wouldn't:
+   *
+   *  - `A with B with C { ... }` is represented with a RefinedType whose
+   *    symbol is a fresh class symbol whose parents are `A`, `B`, `C`.
+   *  - Structural members also get their own symbols.
+   *
+   *  To emulate this, we simply use the type itself a stand-in for its symbol.
+   *
+   *  See also `sameSymbol` which determines if two pseudo-symbols are really the same.
+   */
+  type PseudoSymbol = Symbol | StructuralRef | Scala2RefinedType
+
+  /** The pseudo symbol of `tp`, see `PseudoSymbol`.
+   *
+   *  The pseudo-symbol representation of a given type is chosen such that
+   *  `isNonBottomSubClass` behaves like it would in Scala 2, in particular
+   *  this lets us strip all aliases.
+   */
+  def pseudoSymbol(tp: Type)(using Context): PseudoSymbol = tp.widenDealias match
+    case tpw: Scala2RefinedType =>
+      supportedType(tpw)
+      tpw
+    case tpw: TypeRef =>
+      val sym = tpw.symbol
+      if !sym.exists then
+        // Since we don't have symbols for structural type members we use the
+        // type itself and rely on `sameSymbol` to determine whether two
+        // such types would be represented with the same Scala 2 symbol.
+        tpw
+      else
+        sym
+    case tpw: TypeProxy =>
+      pseudoSymbol(tpw.underlying)
+    case tpw: JavaArrayType =>
+      defn.ArrayClass
+    case tpw: OrType =>
+      pseudoSymbol(TypeErasure.scala2Erasure(tpw))
+    case tpw: ErrorType =>
+      defn.ObjectClass
+    case tpw =>
+      throw new Error(s"Internal error: unhandled class ${tpw.getClass} for type $tpw in pseudoSymbol($tp)")
+
+  extension (psym: PseudoSymbol)(using Context)
+    /** Would these two pseudo-symbols be represented with the same symbol in Scala 2? */
+    def sameSymbol(other: PseudoSymbol): Boolean =
+      // Pattern match on (psym1, psym2) desugared by hand to avoid allocating a tuple
```

Comment:
I think it would be slightly nicer to use pattern matching, even if we needed to use two nested matches for performance reasons. But I leave the final choice to you.

Reply:
I prefer to avoid using two nested matches because then I would need two fallback clauses, so in the end the code wouldn't look clearer than it does currently.